File: Baekjoon_Solve/BruteForce/20208/20208.java
// 진우의 민트초코우유 (20208번)
import java.io.*;
import java.util.*;

public class Main {
	
	static int ans;
	static int N, M, H, L; // 마을 크기, 초기 체력, 체력 증가량, 초코 우유 개수
	static boolean[] visited; // 방문 배열
	static int[] jinwoo; // 진우
	static int[][] graph, pos; // 마을, 모두의 위치
	static ArrayList<int[]> milks; // 우유 리스트
	
	
	static void dfs(int cur, int h, int cnt) {
		// 현재 위치에서 진우네 집으로 다시 올 수 있는지 확인
		int curY = pos[cur][0];
		int curX = pos[cur][1];
		if (h - (Math.abs(curY - pos[0][0]) + Math.abs(curX - pos[0][1])) >= 0) {
			ans = Math.max(ans, cnt);
		}
		for (int i = 0; i < L; i++) {
			int nxtY = pos[i + 1][0];
			int nxtX = pos[i + 1][1];
			int left = h - (Math.abs(curY - nxtY) + Math.abs(curX - nxtX));
			if (!visited[i] && left >= 0) {
				visited[i] = true;
				dfs(i + 1, left + H, cnt + 1);
				visited[i] = false;
			}
		}
	}
	
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		H = Integer.parseInt(st.nextToken());
	
		graph = new int[N][N];
		jinwoo = new int[2];
		milks = new ArrayList<>();
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < N; j++) {
				graph[i][j] = Integer.parseInt(st.nextToken());
				if (graph[i][j] == 1) { // 진우의 위치를 기록
					jinwoo[0] = i;
					jinwoo[1] = j;
				} else if (graph[i][j] == 2) { // 우유의 위치를 기록
					milks.add(new int[] {i, j});
				}
			}
		}
		
		L = milks.size();
		pos = new int[L + 1][2];
		pos[0][0] = jinwoo[0];
		pos[0][1] = jinwoo[1];
		for (int i = 0; i < L; i++) {
			pos[i + 1][0] = milks.get(i)[0];
			pos[i + 1][1] = milks.get(i)[1];
		}
		
		visited = new boolean[L];
		
		ans = 0;
		dfs(0, M, 0);
		
		bw.write(ans + "");
		bw.flush();
		bw.close();
		br.close();
	}
}

File: Baekjoon_Solve/BruteForce/20208/20208.py
import sys
sys.setrecursionlimit(int(1e5))
input = sys.stdin.readline

N, M, H = map(int, input().split())
board = []
arr = [] # 민트초코 좌표를 담을 리스트

for i in range(N):
    board.append(input().rstrip().split())
    for j in range(N):
        if board[i][j] == "2":
            arr.append((i, j))
        elif board[i][j] == "1":
            y, x = i, j

L = len(arr)
visited = [False] * (L) # 중복방문 방지 체크 배열

# 순열을 구성하는 재귀함수
def permute(k, m, cnt, cur_y, cur_x):
    global max_cnt
    if m < 0:
        return
    if k >= 1 and m >= 0:
        if m - (abs(cur_y - y) + abs(cur_x - x)) >= 0:
            max_cnt = max(max_cnt, cnt)
    if k == L:
        return
    for i in range(L):
        if not visited[i]:
            visited[i] = True
            m -= abs(cur_y - arr[i][0]) + abs(cur_x - arr[i][1])
            if m >= 0: # 도중에 0보다 작아지면 민트초코까지 도달할 수 없음
                m += H
                cnt += 1
                permute(k+1, m, cnt, arr[i][0], arr[i][1])
                m -= H
                cnt -= 1
            m += abs(cur_y - arr[i][0]) + abs(cur_x - arr[i][1])
            visited[i] = False

max_cnt = 0
permute(0, M, 0, y, x)
print(max_cnt)